bugfix/postgresql scram passwords

astorath opened the pull request:
##### SUMMARY
Added scram stored passwords check for modification
Fixes #49878

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
postgresql_user

##### ADDITIONAL INFORMATION
I've imported `saslprep` utility function from https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/saslprep.py, don't know how to handle Apache v2 license correctly in this case though.

Commit:
fix(rabbitmq_plugin): fixed an issue, when module successfully reports installation of non existing plugins

Commit:
fix(rabbitmq_plugin): fixed an issue, when module successfully reports installation of non existing plugins

Commit:
feat: added scram stored passwords check for modification

Commit:
Revert "fix(rabbitmq_plugin): fixed an issue, when module successfully reports installation of non existing plugins"

This reverts commit 2c041f0c186bc9eb440e5a5203871e458683ed15.

Commit:
Revert "fix(rabbitmq_plugin): fixed an issue, when module successfully reports installation of non existing plugins"

This reverts commit 2c041f0c186bc9eb440e5a5203871e458683ed15.

Commit:
style(postgresql_user): fixes E303 too many blank lines

Andersson007 commented:
@astorath Please, close and reopen this PR to check ci.

Andersson007 reviewed (changes requested):
SASL Authentication is available from PostgreSQL 10 version. You should add note about this in the documentation section for password option. And add this link to official docs U(https://www.postgresql.org/docs/current/sasl-authentication.html)
Here
```
  41   password:
  42     description:
  43       - Set the user's password, before 1.4 this was required.
  44       - Password can be passed unhashed or hashed (MD5-hashed, SCRAM-SHA-256).
...
```
And maybe better to add one to EXEMPLES with SCRAM-SHA-256 hashed password.

Andersson007 commented:
I tried it - works well. I'm waiting changes from you ;)

Andersson007 commented:
I think you'll have a problem with ci because https://github.com/ansible/ansible/pull/43650, https://github.com/ansible/ansible/pull/51613, https://github.com/ansible/ansible/pull/51220

Andersson007 commented:
@mattclay 

> I've imported `saslprep` utility function from https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/saslprep.py, don't know how to handle Apache v2 license correctly in this case though.

what about this? Is it no problem?

astorath commented:
> And maybe better to add one to EXEMPLES with SCRAM-SHA-256 hashed password.

I'm not sure it is even possible to set SCRAM-SHA-256 hashed password, password should be set plaintext. SCRAM is all about how password is stored and how authentication is made (without passing the password in clear).

I'll look into this and get back with docs.

Andersson007 commented:
we can pass md5 in a hashed form. I tried to pass like
```
passowrd: SCRAM-SHA-256$4096:RmnRFDvK0XWNYCTInV61ww==$WPAaJG/7Kn1BImfxsOYfeZsZ3u4YRgRbnepAYg18LvU=:L0AnLYOUvfUWdho5OlHn/v67EfcpOcjAvjuBNT0TPrU=
```
and then I have:
```
postgres=# select rolpassword from pg_authid where rolname = 'sha_user';
                                                              rolpassword                                                              
---------------------------------------------------------------------------------------------------------------------------------------
 SCRAM-SHA-256$4096:RmnRFDvK0XWNYCTInV61ww==$WPAaJG/7Kn1BImfxsOYfeZsZ3u4YRgRbnepAYg18LvU=:L0AnLYOUvfUWdho5OlHn/v67EfcpOcjAvjuBNT0TPrU=
```
PostgreSQL understand that is a hash and stored it as it is.

abadger commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -244,6 +257,110 @@ class InvalidFlagsError(Exception):
+"""An implementation of RFC4013 SASLprep.
+This code is mostly (apart from utils methods) a copy of
+https://github.com/mongodb/mongo-python-driver/blob/master/pymongo/saslprep.py
+"""
```
I think that licenses for GPLv3+ and Apachev2 allow us to mix this code legally but I'm a little hesitant to say this is the right way to include the code.  It has a separate upstream and a separate license so maybe it would be better to but this into its own file in module_utils and then import it into this code?

That would also make it theoretically possible for us to track updates in the upstream code (in case there were security issues or bugs that we'd want to address, for instance)

Andersson007 replied:
@abadger, I'm not really good at understanding of Ansible internals for a while but code separating sounds reasonably. Besides, in the future, it can be useful for other modules. @astorath can you realized it in own file as a class? (Better in a general reusable form)

astorath replied:
@Andersson007 sorry, I'm in a business trip right now, so I'll finish next week.

Andersson007 commented:
When you finish this PR, this is https://github.com/ansible/ansible/issues/41787 an idea for the next one ;)

Andersson007 commented:
@astorath , waiting for your feedback

Andersson007 commented:
needs_info

Andersson007 commented:
@astorath , hey! hope you're well.
What about this PR? If you don't have enough time to complete it, I can do it by a new one.

astorath commented:
@Andersson007 I've been gone for a while. I'm sorry for that. Will fix this asap.

Andersson007 commented:
@astorath , good news!

Commit:
docs: updated SCRAM-hashed password options in docs, chore: moved saslprep to separate module

Commit:
fix: check for pbkdf2 lib

astorath commented:
@Andersson007, 

I think I've finished finally:
- moved saslprep to a separate module
- updated docs a bit

I've tried to add some tests, but can't manage env setup, it just don't work on my machine, so I ran  tests manually, without any automation.

Andersson007 commented:
@astorath , thank you!
We need integration tests

Andersson007 commented:
@astorath , if there are any difficulties with integration tests I could help you if you don't mind

Commit:
test(postgresql_user): added integration tests for SCRAM-SHA-256

astorath commented:
@Andersson007 ok, I'm done with the tests.

One last thing: I think we should remove setting SCRAM encrypted passwords from examples:
- it's cumbersome (using some kind of third party lib to generate one, set all the params right, then encode the result properly)
- I don't think authors intended this behavior
- it can have compatibility issues (if some parameters will change, e.g. iterations count)
- it's a bit dangerous: you can mess up with salt for instance and make your db vulnerable to attacks

Andersson007 commented on test/integration/targets/postgresql/tasks/test_password.yml:
```diff
@@ -300,6 +300,95 @@
+
+    - name: "Using cleartext password with scram-sha-256: ensure password is properly encrypted"
+      <<: *task_parameters
+      shell: >-
```
FUY there is postgresql_query module. Pretty convenient and not so cumbersome :) than using shell. Look at postgresql_owner.yml, e.g.:
```
 190 # Check, rowcount must be 1
 191 - name: postgresql_owner - check that ownership has been changed after the previous step
 192   become_user: "{{ pg_user }}"
 193   become: yes
 194   postgresql_query:
 195     db: acme
 196     login_user: "{{ pg_user }}"
 197     query: "SELECT 1 FROM pg_tables WHERE tablename = 'my_table' AND tableowner = 'alice'"
 198   ignore_errors: yes
 199   register: result
 200 
 201 - assert:
 202     that:
 203     - result.rowcount == 1
```

astorath replied:
I've fixed that one, but there are plenty of such shell based queries out there...

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -46,16 +46,18 @@
     description:
     - Set the user's password, before 1.4 this was required.
-    - Password can be passed unhashed or hashed (MD5-hashed).
+    - Password can be passed unhashed or hashed (MD5-hashed or SCRAM-SHA256-hashed).
```
Please, add info that scram method is available from 2.9

astorath replied:
Do we have to wait for 2.9? In the end, this PR does not change any interfaces, just fixing reapplying password on SCRAM encrypted installations.

I'll remove this from doc

Andersson007 replied:
@gundalow ?

gundalow replied:
This PR is listed as `Feature Pull Request`, if it's truely a bug fix please update that.

This will not go into 2.8.0, though it *could* go into 2.8.1

astorath replied:
@Andersson007 I've removed this mention, cause it's not a feature (and it's available right now, without any fix)

@gundalow Renamed the PR. May I backport it to 2.8?

gundalow replied:
I've changed this to a bug fix.
1) Please add a changelog/fragment https://docs.ansible.com/ansible/devel/community/development_process.html#making-your-pr-merge-worthy
2) Once this PR has been merged into `devel` you an create a backport by following the process on https://docs.ansible.com/ansible/devel/community/development_process.html#backporting-merged-prs

astorath replied:
done

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -46,16 +46,18 @@
     - Note that if the provided password string is already in MD5-hashed
       format, then it is used as-is, regardless of C(encrypted) parameter.
+    - When passing a SCRAM-hashed password it must be in internal PostgreSQL format,
+      see https://www.postgresql.org/docs/10/catalog-pg-authid.html for reference.
```
```suggestion
      see https://www.postgresql.org/docs/current/catalog-pg-authid.html for reference.
```

astorath replied:
I've removed direct mentions on howto format password manually, but fixed the one in comments

Andersson007 commented on lib/ansible/module_utils/saslprep.py:
```diff
@@ -0,0 +1,119 @@
+
```
```suggestion
```

astorath replied:
typo

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -247,6 +261,16 @@
+try:
+    # pbkdf2_hmac is missing on python 2.6, we can safely assume,
+    # that postresql 10 capable intance have at least python 2.7 installed
+    from hashlib import pbkdf2_hmac
```
```suggestion
    from hashlib import pbkdf2_hmac
    pbkdf2_found = True
```

Andersson007 replied:
then remove 'else'

astorath replied:
ok

Andersson007 commented on lib/ansible/module_utils/saslprep.py:
```diff
@@ -0,0 +1,119 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Maybe your contact here?

astorath replied:
ok

Andersson007 commented on lib/ansible/module_utils/saslprep.py:
```diff
@@ -0,0 +1,119 @@
+        # RFC3454 section 2, step 3 and 4 - Prohibit and check bidi
+        for char in data:
+            if any(in_table(char) for in_table in prohibited):
+                raise ValueError(
```
```suggestion
                raise ValueError("SASLprep: failed prohibited character check")
```

astorath replied:
fixed

Andersson007 commented:
@astorath , thank you
1. A couple of cosmetic suggestions.

2.
>I think we should remove setting SCRAM encrypted passwords from examples

I agree. We can remove it I think.

waiting for your feedback. then I'll approve it

Commit:
style: typos, docs: + contact

Commit:
docs: postgresql docs url, style: pbkdf2_found

Andersson007 commented on test/integration/targets/postgresql/tasks/main.yml:
```diff
@@ -238,6 +238,10 @@
     encryption_values: '{{ encryption_values }} + ["no"]'
   when: postgres_version_resp.stdout is version('10', '<=')
 
+# scram-sha-256 password encryption type is supported from PostgreSQL 10.0
```
I think better to remove it
you can use intro test_password.yml a condition like this
```
when: postgres_version_resp.stdout is version('10', '>=')
```

Andersson007 replied:
or define this variable into there

Andersson007 replied:
better to keep main.yml cleaner (better to use it only for includes)

astorath replied:
agreed

Commit:
style: rolled back changes to main.yml in tests,  style: updated db query task to new postgresql_query module

Andersson007 commented:
@astorath thank you! Please, mark our conversations as solved (where you solved them :)

and we are waiting feedback about backporting from @gundalow 
and about license from @Akasurde

astorath commented:
> @astorath thank you! Please, mark our conversations as solved (where you solved them :)

@Andersson007 I've resolved everything but 2 concerning docs, cause I need input about backporting first.

Commit:
docs: removed setting SCRAM enced password mentions

Andersson007 reviewed (approved):
shipit

Andersson007 commented:
Oh, @astorath , add a changelog fragment please as gundalow wrote

@gundalow , do you know who can help us with the licenses compatibility ?
> akasurde> Abhijeet Kasurde andersson007_, lib/ansible/module_utils/saslprep.py used in PR is using Apache-2 license
1:33 PM not sure if that is compatible with BSD license

Commit:
docs: updated changelog

Andersson007 commented:
!needs_revision

shipit

Commit:
docs: added issue ref

Andersson007 commented:
shipit

Andersson007 commented:
@kostiantyn-nemchenko could you please look at this one?

kostiantyn-nemchenko commented:
@Andersson007 Will try to take a look this week.

kostiantyn-nemchenko commented:
shipit

Andersson007 commented:
>5:43 PM <bcoca> must be bsd for module_utils, gpl for rest

@astorath , problem with licensing. could you change it please?

Commit:
docs(saslprep): changed license header to GPLv3

astorath commented:
> @astorath , problem with licensing. could you change it please?

Moved to GPLv3, as far as I've learned so far, this is legitimate.

sivel commented on lib/ansible/module_utils/saslprep.py:
```diff
@@ -1,16 +1,5 @@
-# See the License for the specific language governing permissions and
-# limitations under the License.
+# Copyright (c) 2019 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
This code is copyright to MongoDB, Inc.  I don't think you can simply switch out copyright and license like this.

My recommendation on IRC was that `pymongo` should just be made a soft dependency of the module, and only required as needed.

astorath replied:
@sivel Yes, I should have kept the copyright.

I don't want to add `pymongo` as a dependency, this makes no sense in this case.

1. I can keep the copyright and use dual license in this case (it is permitted by Apache2 / GPL3 AFAIK) 
2. I can use another implementation licensed under BSDm keeping the copyright and replacing the license (it is permitted by BSD / GPL3 AFAIK)
3. I can rewrite this one from scratch...

My personal preference is (2) if (1) is not an option. Are you ok with (1) or (2)?

astorath replied:
@Andersson007

Andersson007 replied:
hey:) I saw and tried to ask @sivel in irc for looking at this yesterday. Probably we should try to do it again today later.
(In my opinion, if I could I tried to rewrite this actually from scratch with a suitable license)

sivel replied:
Use of any license other than BSD for `module_utils` is going to require at minimum core team agreement.  As far as dual licensing is concerned, we do not accept dual licenses.  You would need to convince the original authors to accept a new license.  Any work related to this is likely to require discussions with the Red Hat legal team as well.

In short, our recommendation would be to find another bit of code that is BSD licensed, or to not bundle this code and require a new external dependency for modules needing this functionality.

I think your preference for the 2nd option would be preferable. Finding another implementation already licensed under BSD.

andytom replied:
We could look at using [Passlib](https://passlib.readthedocs.io/en/stable/) it is BSD licensed.

astorath replied:
> We could look at using Passlib it is BSD licensed.

@andytom @sivel
Yep, I was going to take passlib implementation, it needs some work, but it's fine, thanks!

astorath replied:
@sivel 
I've updated `saslprep.py` module sources (now under BSD license).

Commit:
chore: replaced saslprep impl with BSD licensed one

Commit:
test: added saslprep tests

Andersson007 commented:
@astorath , i sent several direct emails to you but I didn’t get any answers.

needs_info

kravietz commented:
@Andersson007   If the original author @astorath is busy and unable to finish this PR, I can complete it as I actually wanted to submit fix for this on my own.

Andersson007 commented:
@kravietz , would be really good!

astorath commented:
@Andersson007 , was very busy recently, sorry. I've merged develop branch once again, have some trouble with integration tests (they changed a lot since last time I've used them). 

I hope to update the PR this week to proceed with merge.

Andersson007 commented:
@astorath np, understand
@kravietz , you told us about some fixes of this PR
could you please share it with @astorath ?
Would be great to see you both working on this patch together

Andersson007 commented:
@astorath , FYI integration tests were moved from ```postgres``` directory to separate targets. For postgresql_user to ```test/integration/targets/postgresql_user``` respectively

astorath commented:
@Andersson007 , yep, I saw that. The problem is, fedora28 image was replaced with fedora29, and the last one has some strange `tar` error on ansible setup stage...

All the tests are fine now, I expect to update the PR today after some review.

Commit:
cleanup: removed rebase artifacts

Commit:
chore: fixing sanity metaclass-boilerplate

astorath commented:
@Andersson007 @sivel I've merged devel once again and fixed issues. Please review

Commit:
chore: fixing sanity metaclass-boilerplate

Neustradamus commented:
Any news?

Andersson007 commented:
@Neustradamus it looks like it is free for fixing now:) the work stumbled on the licencing.. would be really cool if anybody complete this PR (but in https://github.com/ansible-collections/community.general since all not core stuff have been recently moved to there)

Neustradamus commented:
@tinproject @astorath: Have you seen the comment of @Andersson007?

Andersson007 commented:
if nobody is in progress of solving this, i've just started rewriting

Andersson007 commented:
Who's interested in this feature is welcome to review:) https://github.com/ansible-collections/community.general/pull/100
Despite CI, would be cool if anybody try it manually too.

Andersson007 commented:
close_me

Andersson007 commented:
closed via https://github.com/ansible-collections/community.general/pull/100
